### PR TITLE
Allow negative scale

### PR DESCRIPTION
**Pull request by BenjaminHug8**
**Description**
In order to support the negative scale on meshes in `URDF` files, this PR adds the possibility to specify negative `scale` in the `Transform` node and to specify a `ccw` flag to revert the faces for `Mesh` and `ColladaShapes` nodes.

### CODE REVIEW

File: src/webots/nodes/WbAbstractTransform.cpp [hunk 1]
```diff
@@ -109,44 +109,29 @@ void WbAbstractTransform::updateTranslationAndRotation() {
...
-    if (x == 0.0) {
-      correctedScale.setX(1.0);
-      mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be positive: x is set to 1.0."));
-    } else {
-      correctedScale.setX(fabs(x));
-      mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be positive: x is set to abs(x)."));
-    }
+  if (x == 0.0) {
+    correctedScale.setX(1.0);
+    mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be not null: x is set to 1.0."));
```
Comment (omichel) [inline src/webots/nodes/WbAbstractTransform.cpp:122]:
```suggestion
    mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be non-zero: x is set to 1.0."));
```

File: src/webots/nodes/WbAbstractTransform.cpp [hunk 2]
```diff
@@ -109,44 +109,29 @@ void WbAbstractTransform::updateTranslationAndRotation() {
...
-    if (y == 0.0) {
-      correctedScale.setY(1.0);
-      mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be positive: y is set to 1.0."));
-    } else {
-      correctedScale.setY(fabs(y));
-      mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be positive: y is set to abs(y)."));
-    }
+  if (y == 0.0) {
+    correctedScale.setY(1.0);
+    mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be not null: y is set to 1.0."));
```
Comment (omichel) [inline src/webots/nodes/WbAbstractTransform.cpp:128]:
```suggestion
    mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be non-zero: y is set to 1.0."));
```

File: src/webots/nodes/WbAbstractTransform.cpp [hunk 3]
```diff
@@ -109,44 +109,29 @@ void WbAbstractTransform::updateTranslationAndRotation() {
...
-    if (z == 0.0) {
-      correctedScale.setZ(1.0);
-      mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be positive: z is set to 1.0."));
-    } else {
-      correctedScale.setZ(fabs(z));
-      mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be positive: z is set to abs(z)."));
-    }
+  if (z == 0.0) {
+    correctedScale.setZ(1.0);
+    mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be not null: z is set to 1.0."));
```
Comment (omichel) [inline src/webots/nodes/WbAbstractTransform.cpp:134]:
```suggestion
    mBaseNode->parsingWarn(QObject::tr("All 'scale' coordinates must be non-zero: z is set to 1.0."));
```

File: src/webots/nodes/WbAbstractTransform.cpp [hunk 4]
```diff
@@ -109,44 +109,29 @@ void WbAbstractTransform::updateTranslationAndRotation() {
     updateTranslateRotateHandlesSize();
 }
 
-bool WbAbstractTransform::checkScalePositivity(WbVector3 &correctedScale) const {
+bool WbAbstractTransform::checkScaleNullValues(WbVector3 &correctedScale) const {
```
Comment (omichel) [inline src/webots/nodes/WbAbstractTransform.cpp:112]:
```suggestion
bool WbAbstractTransform::checkScaleZeroValues(WbVector3 &correctedScale) const {
```

File: src/webots/nodes/WbAbstractTransform.cpp [hunk 5]
```diff
@@ -379,7 +364,7 @@ void WbAbstractTransform::updateMatrix() const {
   assert(mMatrix);
 
   mMatrix->fromVrml(mTranslation->x(), mTranslation->y(), mTranslation->z(), mRotation->x(), mRotation->y(), mRotation->z(),
-                    mRotation->angle(), mScale->x(), mScale->y(), mScale->z());
+                    mRotation->angle(), abs(mScale->x()), abs(mScale->y()), abs(mScale->z()));
```
Comment (omichel) [inline src/webots/nodes/WbAbstractTransform.cpp:367]: That seems wrong... Do we really need that?

File: src/webots/nodes/WbAbstractTransform.hpp [hunk 6]
```diff
@@ -147,7 +147,7 @@ class WbAbstractTransform {
   // A specific scale check is done in the WbSolid class
   WbSFVector3 *mScale;
   bool checkScale(int constraintType = 0, bool warning = false);
-  bool checkScalePositivity(WbVector3 &correctedScale) const;
+  bool checkScaleNullValues(WbVector3 &correctedScale) const;
```
Comment (omichel) [inline src/webots/nodes/WbAbstractTransform.hpp:150]:
```suggestion
  bool checkScaleZeroValues(WbVector3 &correctedScale) const;
```

File: src/webots/nodes/WbTriangleMeshGeometry.cpp [hunk 7]
```diff
@@ -529,7 +538,10 @@ void WbTriangleMeshGeometry::updateNormalsRepresentation() {
     QVector<float> vertices;
     QVector<float> colors;
     const int n = mTriangleMesh->numberOfTriangles();
-    const double linescale = WbWorld::instance()->worldInfo()->lineScale();
+    int orientation = 1;
+    if (!mCcw)
+      orientation = -1;
```
Comment (omichel) [inline src/webots/nodes/WbTriangleMeshGeometry.cpp:543]:
```suggestion
    const int orientation = mCcw ? 1 : -1;
```

File: projects/shapes/protos/ColladaShapes.proto [hunk 8]
```diff
@@ -14,12 +15,17 @@ PROTO ColladaShapes [
       console.error(`The given Collada url (${url}) is not valid.`);
       url = fields.url.defaultValue;
     }
+
+    let ccwString = "TRUE";
+    if (!fields.ccw.value) {
+      ccwString = "FALSE";
+    }
```
Comment (omichel) [inline projects/shapes/protos/ColladaShapes.proto:22]:
```suggestion
    const ccwString = (fields.ccw.value) ? "TRUE" : "FALSE";
```

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you.
That looks good to me.
I am just not sure about the change in WbAbstractTransform::updateMatrix which seems wrong to me...

File: src/webots/gui/WbDragTransformEvent.cpp [hunk 9]
```diff
@@ -215,8 +215,8 @@ void WbDragTranslateAlongAxisEvent::apply(const QPoint &currentMousePosition) {
   mViewDistanceUnscaling = mViewpoint->viewDistanceUnscaling(mSelectedTransform->position());
 
   WbMatrix4 matrix(mSelectedTransform->matrix());
-  const WbVector3 absoluteScale = matrix.scale();
-  matrix.scale(1.0f / absoluteScale.x(), 1.0f / absoluteScale.y(), 1.0f / absoluteScale.z());
+  WbVector3 scale = mSelectedTransform->scale();
```
Comment (omichel) [inline src/webots/gui/WbDragTransformEvent.cpp:218]: Can't we use const here?
```suggestion
  const WbVector3 scale = mSelectedTransform->scale();
```
Or even better:
```suggestion
  const WbVector3 &scale = mSelectedTransform->scale();
```

File: src/webots/gui/WbDragTransformEvent.cpp [hunk 10]
```diff
@@ -173,8 +173,8 @@ WbDragTranslateAlongAxisEvent::WbDragTranslateAlongAxisEvent(const QPoint &initi
   mTextOverlay->applyChangesToWren();
 
   WbMatrix4 matrix(mSelectedTransform->matrix());
-  const WbVector3 absoluteScale = matrix.scale();
-  matrix.scale(1.0f / absoluteScale.x(), 1.0f / absoluteScale.y(), 1.0f / absoluteScale.z());
+  WbVector3 scale = mSelectedTransform->scale();
```
Comment (omichel) [inline src/webots/gui/WbDragTransformEvent.cpp:176]: Can't we use `const` here?

File: src/webots/wren/WbWrenAbstractResizeManipulator.cpp [hunk 11]
```diff
@@ -136,6 +136,7 @@ void WbWrenAbstractResizeManipulator::initializeHandlesEntities(bool resize) {
 
   for (int i = 0; i < mNumberOfHandles; ++i) {
     WrRenderable *renderable = wr_renderable_new();
+    wr_renderable_set_scene_culling(renderable, false);
```
Comment (omichel) [inline src/webots/wren/WbWrenAbstractResizeManipulator.cpp:139]: Why is this needed?

File: src/wren/Renderable.cpp [hunk 12]
```diff
@@ -279,6 +292,9 @@ void wr_renderable_set_drawing_mode(WrRenderable *renderable, WrRenderableDrawin
 void wr_renderable_set_visibility_flags(WrRenderable *renderable, int flags) {
   reinterpret_cast<wren::Renderable *>(renderable)->setVisibilityFlags(flags);
 }
+void wr_renderable_invert_front_face(WrRenderable *renderable, bool invert_front_face) {
```
Comment (omichel) [inline src/wren/Renderable.cpp:295]:
```suggestion
}

void wr_renderable_invert_front_face(WrRenderable *renderable, bool invert_front_face) {
```

File: src/wren/Renderable.cpp [hunk 13]
```diff
@@ -219,8 +220,20 @@ namespace wren {
     glUniformMatrix4fv(program->uniformLocation(WR_GLSL_LAYOUT_UNIFORM_MODEL_TRANSFORM), 1, false,
                        glm::value_ptr(mParent->matrix()));
 
+    // to render cw and ccw meshes
+    const unsigned int frontFaceMode = glstate::getFrontFace();
+    if (mInvertFrontFace) {
+      if (glstate::getFrontFace() == GL_CCW)
```
Comment (omichel) [inline src/wren/Renderable.cpp:226]:
```suggestion
      if (frontFaceMode == GL_CCW)
```

**Review by omichel — CHANGES_REQUESTED** [review]
Nice job. It looks pretty good.

File: src/webots/gui/WbDragTransformEvent.cpp (hunk 10 above)
Comment (BenjaminHug8) [reply src/webots/gui/WbDragTransformEvent.cpp:176]: I did it originally but Cpp-check was complaining...

Comment (BenjaminHug8) [reply src/webots/gui/WbDragTransformEvent.cpp:176]: Ok I misunderstood, in facts it was asking for a const reference, as you suggested.

File: src/webots/wren/WbWrenAbstractResizeManipulator.cpp (hunk 11 above)
Comment (BenjaminHug8) [reply src/webots/wren/WbWrenAbstractResizeManipulator.cpp:139]: It is for consistency with the transform manipulator.

**Comment by BenjaminHug8** [comment]
The `wr_renderable_set_scene_culling` is used for manipulators in case the axes system is close to the edges of the window. Without it it might disappear although it should be displayed.

File: resources/nodes/Mesh.wrl [hunk 14]
```diff
@@ -1,7 +1,8 @@
 # The Mesh node represents a 3D geometry defined in an external file.
 
 Mesh {
-  field MFString url [ ]   # URL of a 3D file such as DAE, STL, OBJ or FBX format (similar to ImageTexture node)
-  field SFString name ""   # Filter geometries based on the given name (for example, Collada files can contain multiple geometries in a single file).
-  field SFInt32 materialIndex -1   # Filter geometries in Collada files based on the material index (Collada files can contain multiple geometries with different materials).
+  field MFString url [ ]         # URL of a 3D file such as DAE, STL, OBJ or FBX format (similar to ImageTexture node)
+  field SFBool ccw TRUE          # Indicate the order of the vertices. Changing this value will flip the faces of the mesh.
+  field SFString name ""         # Filter geometries based on the given name (for example, Collada files can contain multiple geometries in a single file).
+  field SFInt32 materialIndex -1 # Filter geometries in Collada files based on the material index (Collada files can contain multiple geometries with different materials).
```
Comment (omichel) [inline resources/nodes/Mesh.wrl:7]: We should align fields and values to be consistent with other nodes.
```suggestion
  field MFString url           []   # URL of a 3D file such as DAE, STL, OBJ or FBX format (similar to ImageTexture node)
  field SFBool   ccw           TRUE # Indicate the order of the vertices. Changing this value will flip the faces of the mesh.
  field SFString name          ""   # Filter geometries based on the given name (for example, Collada files can contain multiple geometries in a single file).
  field SFInt32  materialIndex -1   # Filter geometries in Collada files based on the material index (Collada files can contain multiple geometries with different materials).
```

File: src/wren/Renderable.cpp [hunk 15]
```diff
@@ -219,8 +220,20 @@ namespace wren {
...
                        glm::value_ptr(mParent->matrix()));
 
+    // to render cw and ccw meshes
+    const unsigned int frontFaceMode = glstate::getFrontFace();
+    if (mInvertFrontFace) {
+      if (frontFaceMode == GL_CCW)
+        glstate::setFrontFace(GL_CW);
+      else
+        glstate::setFrontFace(GL_CCW);
+    }
```
Comment (omichel) [inline src/wren/Renderable.cpp:230]: Simpler this way:
```suggestion
    if (mInvertFrontFace)
      glstate::setFrontFace((frontFaceMode == GL_CCW) ? GL_CW : GL_CCW);
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by omichel — APPROVED** [review]
Thank you.
